### DISCUSSION TITLE
Adding models to the Glossary - 0

### DISCUSSION THREAD

**Discussion by jsvgoncalves** [Depart-de-Sentier/sentier.dev#24] [Workshop Item]
**Goal**: Include models that generate LCI datasets in the glossary and link them to other glossary concepts.

**Key Questions:**
- **Model Schema:** How do we represent a model in the glossary schema? What is the minimum amount of information we need to store about a model? 
- **Model Execution:** How do we pipe data from the glossary to the model? What sort of facade or adapter is need to map the model's IO to the glossary?
- **Data Uncertainty:** For a given execution flow, where is information about uncertainty given? Is it a pre-requirement for model inclusion in the glossary? 

**Potential Outcomes:**
- Addition of `Models` as an entity type to the glossary.
- Definition and implementation of first mini-pipeline using the glossary.

**Candidate Models:**
- [https://github.com/msmasnadi/OPGEEv4](https://github.com/msmasnadi/OPGEEv4)
- [https://onlinelibrary.wiley.com/doi/abs/10.1002/pip.3695](https://onlinelibrary.wiley.com/doi/abs/10.1002/pip.3695)
- [https://github.com/romainsacchi/LCA_WIND_DK/tree/master](https://github.com/romainsacchi/LCA_WIND_DK/tree/master)
- [https://github.com/romainsacchi/carculator](https://github.com/romainsacchi/carculator)
- [https://pubs.acs.org/doi/10.1021/acs.est.3c03190](https://pubs.acs.org/doi/10.1021/acs.est.3c03190)
- [https://www.epa.gov/warm/versions-waste-reduction-model#v16](https://www.epa.gov/warm/versions-waste-reduction-model#v16)
-  [https://www.sciencedirect.com/science/article/abs/pii/S0360132323011046](https://www.sciencedirect.com/science/article/abs/pii/S0360132323011046)

**Comment by jsvgoncalves** [comment]
Here's the list of current schemes we're importing: 
https://github.com/sentier-dev/glossary-api/blob/2be9056ede31c2de4839e18875e5bd4376ab2c5b/dds_glossary/services.py#L50-L93

```python
class GlossaryController:
    """
    Controller for the glossary.

    Attributes:
        engine (Engine): The database engine.
        data_dir (Path): The data directory for saving the datasets.
    """

    europa_url: ClassVar[str] = "http://publications.europa.eu/resource/distribution/"
    fao_url: ClassVar[str] = (
        "https://storage.googleapis.com/fao-datalab-caliper/Downloads/"
    )
    datasets: ClassVar[list[Dataset]] = [
        Dataset(
            name="ESTAT-CN2024.rdf",
            url=(
                europa_url
                + "combined-nomenclature-2024/20240425-0/rdf/skos_core/ESTAT-CN2024.rdf"
            ),
        ),
        Dataset(
            name="ESTAT-LoW2015.rdf",
            url=europa_url + "low2015/20240425-0/rdf/skos_core/ESTAT-LoW2015.rdf",
        ),
        Dataset(
            name="ESTAT-NACE2.1.rdf",
            url=(europa_url + "nace2.1/20240425-0/rdf/skos_core/ESTAT-NACE2.1.rdf"),
        ),
        Dataset(
            name="ESTAT-ICST-COM.rdf",
            url=(europa_url + "icst-com/20240425-0/rdf/skos_core/ESTAT-ICST-COM.rdf"),
        ),
        Dataset(
            name="ESTAT-PRODCOM2023.rdf",
            url=(
                europa_url
                + "prodcom2023/20240425-0/rdf/skos_core/ESTAT-PRODCOM2023.rdf"
            ),
        ),
        Dataset(name="ISIC4.rdf", url=fao_url + "ISICRev4/ISIC4.rdf"),
        Dataset(name="ICC11.rdf", url=fao_url + "ICCv1.1/ICC11.rdf"),
        Dataset(name="WCACROPS.rdf", url=fao_url + "WCA2020Crops/WCACROPS.rdf"),
    ]

**Comment by simontaurus** [comment]
Data schema drafts

[Entity](https://oo-ld.github.io/playground-yaml/index.html?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3FoQUgCZFCyqFAgoA9HYpOAdBQDMz6QHM7ARgCcgXYADABMdqGhALRmRFGIgrpk%2BlFIAMQKscgoIFY29o4u7p5SPqHBFSG%2BdrHxKmpkGWxgFPI5ebYOThSuHt4RFeHBAOx2LdhNigDWYNm51hCIBd29JT4AKgBKdogziDFURFRSVPAYVHYKpjzHdHMdBRAiUthrdtd2ejKIl2xKqup2gslg5%2Fg03gpEKdYMdLMCCmAwFFsKZnF4YFBlCI8M44HYvNhZogyFJBFEyBAYL8QFBBBAqIRNLT6QoYFIYIyYKYFAI8FRsNktFzCLtZigICciAAZMh8%2BRsbSkAxGUB6QyMgBuchEDLYAngXkZ%2Bq8IjIXgZAF8LWwNcdEHB4IRxihbVJ7QgAJLwEgKCVUDVwESIXiu92OnLOiVwKQANTtDquYAwIgYSVgCEIWRQUKoAEEBWBIaIpOdM%2BYljBEAARQ4weBUUwAIUEPMwp1LOWuXzuKFw5yQupAZDEymkgpAmmFnZu3xQGBOZCg0gUit0yo0IHgZAYIvqshQW53VutNIx2AZOQAoroMS22MyL%2BwuDxMlQAI4iNkN1AAbRAvP5ZAAF02AlSBjhhccHw3aCckQKB2QNK5DmHbAjEUaEYGOFBr1gWkQBPNkOWMWDFAQutDXvM9Hw9TYPQIvVZUAmC6UfEkpCEBQSBkRdCHIPBzwUPDBJyGU5QACkQABKZClHZaAE1EpjsEkqSAAIwCINTMSoNTMPwtgMTUKDWKdThuHEKioBEkAxP5X0pHAqRIJg9cSNMuDyKQtg6Drc8DUxVBfBPY0WJZTzEMokBThTX9osdNhTEHIgpBAEDwAUhBxzrJ4%2BAoLlApyXwKggfQCOPRKUJENDf1VNyQAAVXgQ9v0YpCclOAj0t8%2BAPT0O4gpPUNFNVDyyMioTqMIYaM0qohUPQ3xnGCZaGJAP0A1EYMZvDUbwvGijJusx8AAUTk2oM1LjN0HTU2j6JPUwkxTaFFxGmkxvgiarJsqtntTch00dE9aGeDs9sfL7Dp%2Bx8AGVi3ONS7D0wxji3bA1KSog6wxBMQrbM4LwhkUvKimK6DijoQTsecqDe20sDoOgsreJ4BMwKIntkOsaYwYJqSpgpafp7h6GZpA3j7U4oR%2BPBBDsAAWZbLnSyAgfHcneGEmg4oAYV14IFH1qJGwATTSq02GHTExzCtipA4u9iGkWR0P4wSYcIXMR1twz%2BvHMD6WcrDxxau2SYm48KsUUcKF4Y5HLdQg6y%2BMgeATe8AUfPBESgL6KQVhQyA4ZJUAQnU2CmKgqAgXgECoTSQ21HWUHmgUqBPDhaF2j79qhG9sNw29WXZQgeWU7Ifzq9UcialruTaw0OuB9KdsIJaVsN0DzsDbb4wzHJE2TAG3oPkAizBx9W37Dt5hsanhdgBmxZZ%2Fo2b7Tn6DIHmMD56lrdHEnFAP4gJWiAA%3D%3D%3D)

[Process](https://oo-ld.github.io/playground-yaml/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NsGAC02ACYA6AOYwoykXjdwNh7YYIiIZFKCjmQQMDZKqrI2AKK63oJuiCF8HLRIKmpkbrkIrKB4COEwhOaWtjZSLkRuFUhkMG4ikjZkPDAAbulQghBUNgZweGSIVADMvPMAjADsNgow8BBiyGggmlJU%2FVRSM4StVSgbW1AA8kQKmsOjhJowLiAAvmyiltsaewcjicqGdKu0UD9rncHk8QSg9m9Ph8ALpsOTYaEmUAAEgO93htSs9QczncXh8fgCYCCITCESiMTiCUKKTSw0y2V4JXyiSK3M%2BqJA2KwdAYukIAEEAAQABSkYAwNEQUpgyrI8ClqVgwxYUtoUuoKroEGkUClPioUp0ekMUoAFKMpFLtKQDFAAJRS9UuPUFWT2x1e7AY5q4qhET1gIiatmCBTa7BwkByhVK%2BMjJNgTjccRsA4ARxEMAO7xMIFhIEFEHljtgNH%2BFd2wvoYqMIAl2AOZBcgilLnDGyoPo2eqoaWOKC16SljgAfAbLWAgVI3paOCJEGaLX3w2QRNhcyB%2B0Q9wfCDNx1IUCnFWFPmxi9VdsfT22L3WrwBZQQ3tNfEBXH8uyNiAERSEICgkDIZBtuQeCJgo%2FZKCu0BwPAhA3PAi5OnQ%2B6wBAiZStWYAuCIPDIGwxG1oINyNMcqAAEwAAwsQ%2Beh0DsoAgZuK7wB48beAh8IpqR5EKM2opjm2qQiHQUrrpuUpQeaNCwHxChjrJqCmO%2BMATgAynoVDyGwukTgAwggGAHHolYfP%2BkJAVxGaEGBEFsFBsiwWQ8EgmwSHWTAqGlPCmHYVKuEHkFhHEaJUAUSAVHHMMtH9lIjEsYsbFqJx5YufCPEbPxbAJkmIlkfF4kiq2hAyXJClmspeibkVGnwFpZZmVeADqGwuAA4mOxwwdIChdSg3XTLZyL2fZpnKGAFC8Mc8onIQGx6OBfSlCViRJhUYDxVA4EQAALAoZAcGQ%2BioMdIhUGwADWVBUBAvAIFQUa8P0cj3TsJ7YDM%2F78sB%2BWKFJelXr%2Bd4Piu54QxO37QwlgGVSYgqOWjxiCgIeDGTsxhcW6hAAKrwPAZAMO8bACOp8JjgK%2FkKrJUkwWhhBIkAA%3D)

[Product](https://oo-ld.github.io/playground-yaml/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NsGAC02ACYA6AOYwoykXjdwNh7YYIiIZFKCjmQQMDZKqrI2AKK63oJuiCF8HLRIKmpkbrkIrKB4COEwhOaWtjZSLkRuFUhkMG4ikjZkPDAAbulQghBUNgZweGSIVADMvPMAjADsNgow8BBiyGggmlJU%2FVRSM4StVSgbW1AA8kQKmsOjhJowLiAAvmyiltsaewcjicqGdKu0UD9rncHk8QSg9m9Ph8ALpsOTYaEmUAAEgO93htSs9QczncXh8fgCYCCITCESiMTiCUKKTSw0y2V4JXyiSK3M%2BqJA2KwdAYukIAEEAAQABSkYBcIh4UpgiClZHgUtSsGGLCltCl1BVdAg0igUp8VClOj0hilAApRlIpdpSAYoABKdXwFz6gqyB1O9XYDHNXFUIhesBELVswQKHXYOEgOUKpXiNiwwhgTjcDMgA4ARxEMAO7xMICzgog8qdsBo%2Fyzu2F9DFRhAEuwBzILkEUpcEY2VF9G31VDSxxQ2vSUscAD5DVawECpG8rRwRIhzZb%2BxGyCJsPmB0R94fCDMJ1IUKnFTxPmxS9VdsfT%2B2L%2FWrwBZQQ39NIr6KMoYAULwxzyichAbHoUi9LApSZokyYVGAUBbjBEAACwKGQHBkPoqBQFIIhUGwADWVBUBAvAIFQ0a8P0cjETsJ7YDMAH8rsTaKOOH7XvKt75o%2B548TAk7fr%2Bd5sAIeBUKxqDGKAtrtgAqvA8BkAw7xSRqHiEOOApsC4YAYCIbZkHB8CEEiQA%3D%3D%3D)

[ConceptSchema](https://oo-ld.github.io/playground-yaml/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NsGAC02ACYA6AOYwoykXjdwNh7YYIiIZFKCjmQQMDZKqrI2AKK63oJuiCF8HLRIKmpkbrkIrKAMdHhUUsgoIC481XTWymSIAAoR4gC%2BALpsctgA8kSoxqAAJFJUo3XmltZ2Ds7uXj5%2BAWBBIWERUTFxCYUpaVAZWWA5eUeyxXkgfWwTWHQMuoQAggAEAMIIGFRoABlAqyL4wRBfMjwL6pWBnFhfWhfajgugQaRQL4%2BKhfHR6QxfAAUEGqX20pAMUAAlFD4C4kaCyMTSVIodhhkQ3FMZrSwERYadBAp4dgqIQ%2FvAAcCmSLBKTCGBONxxGxpgBHEQwaYuMYgM4K%2FogCBSSDVWA0DT6%2BXitAgZ70N5GEAfbDTMguQRfFwzGDwKgMv1IqhpaooOHpL6OAB8KNxYAAbtUpDAfV8OCJEFicd6ZmQRNhVfU8wXnYgQxapChJdKoCDEmQHmxtTArT6iPnC4Ry6GqwBZQQ1wF12XdNgVKo1K0G20gCJSIQi7xiwh9tST5BsE1mqRnIZSH1SVAAJgADOfm3pmtObd2oCn4B4l1AV3Uh9AHt0v2xEMowBReGTaRahAP09AXHg4HgEVElnPAHCgLMFwgAAWBQyA4Mh9FQe8RCoNgAGsqEBXgECofleATOQ8NqDtsHLMcQBKaC7RnbsKxgMN3xHBsFBbdjexQAduPrQoFAnapamMI0BCqeixlAAlnQAVXgeAyAYXU2AER9CBDB4jSTGooMIABGNxTwshQTSoBM4EzSjJJMuoFBcMAMBEJ0yFgUoXJ%2FUQpABQgFFwAEpVneYrFsGwMHdWAkxeOgEEQNxpA8GwID8ULHDc2Q%2FRijBTxsdCxD%2FKcTF6L8gA%3D%3D)

[Concept](https://oo-ld.github.io/playground-yaml/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NsGAC02ACYA6AOYwoykXjdwNh7YYIiIZFKCjmQQMDZKqrI2AKK63oJuiCF8HLRIKmpkbrkIIAC6bHLYAPJEqMagACRSVHUoZhZWtvZOrp7evv6BwaHhkdGx8QVJqbBQGVlgOXkJhcV5IAC%2BFSCNWHQMuoQAggAEAMIIGFTQpzCIp2Twp7PpLKe0p9R3dBDSUKcfFRTjo9IZTgAKCBUKSnbSkAxQACUj3gLg%2B0zIkOhsKqtTczVaKLARBeaXmCjm2CohEu8Gu0EpgmhhDAnG44jYLQAjiIYC0XPUQPMWTsIFJIDDYDQNMLmTS0Lt9ocjCBjtgWmQXIJTi5WjB4FR0QaPlQ0jCUK95qdHAA%2BL7AsAANxhUhgetOHBEiABQN1rTIImwnJAeqIgeDhEQZulUhQdIZ4k2bH5MFlYYjquj5rjCZuSc2ycUyjAFF4rukyHaBr0UjIPDg8EpiQVIDwDigPrrEAALAoyBwyPpUFApCIqGwANZUG68BBUEm8J1ycdV8PYaNFkpNxUi1vZ2Pxq75hSpqMxmAWvOMtgCPBUDf1UBg1UAVXg8DIDEFt6eHkIZpbDsLpSIgjaEAAjG4AAMMEKOKVBOnA3pLjCYGlO0CguGAGAiCqZCwBhIAKLQIhSNchAKLg1z0q25iWNYdgYJqsAusqCCIG40geDYEB%2BNRjjYbIBo2BgGDQTY%2FZiCWoH1GUhZAA%3D%3D%3D)

Python Codegeneration:
[Process Dataclass](https://oo-ld.github.io/playground-python-yaml/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NsGAC02ACYA6AOYwoykXjdwNh7YYIiIZFKCjmQQMDZKqrI2AKK63oJuiCF8HLRIKmpkbrkIrKB4COEwhOaWtjZSLkRuFUhkMG4ikjZkPDAAbulQghBUNgZweGSIVADMvPMAjADsNgow8BBiyGggmlJU%2FVRSM4StVSgbW1AA8kQKmsOjhJowLiAAvmyiltsaewcjicqGdKu0UD9rncHk8QSg9m9Ph8ALpsOTYaEmUAAEgO93htSs9QczncXh8fgCYCCITCESiMTiCUKKTSw0y2V4JXyiSK3M%2BqJA2KwdAYukIAEEAAQABSkYAwNEQUpgyrI8ClqVgwxYUtoUuoKroEGkUClPioUp0ekMUoAFKMpFLtKQDFAAJRS9UuPUFWT2x1e7AY5q4qhET1gIiatmCBTa7BwkByhVK%2BMjJNgTjccRsA4ARxEMAO7xMIFhIEFEHljtgNH%2BFd2wvoYqMIAl2AOZBcgilLnDGyoPo2eqoaWOKC16SljgAfAbLWAgVI3paOCJEGaLX3w2QRNhcyB%2B0Q9wfCDNx1IUCnFWFPmxi9VdsfT22L3WrwBZQQ3tNfEBXH8uyNiAERSEICgkDIZBtuQeCJgo%2FZKCu0BwPAhA3PAi5OnQ%2B6wBAiZStWYAuCIPDIGwxG1oINyNMcqAAEwAAwsQ%2Beh0DsoAgZuK7wB48beAh8IpqR5EKM2opjm2qQiHQUrrpuUpQeaNCwHxChjrJqCmO%2BMATgAynoVDyGwukTgAwggGAHHolYfP%2BkJAVxGaEGBEFsFBsiwWQ8EgmwSHWTAqGlPCmHYVKuEHkFhHEaJUAUSAVHHMMtH9lIjEsYsbFqJx5YufCPEbPxbAJkmIlkfF4kiq2hAyXJClmspeibkVGnwFpZZmVeADqGwuAA4mOxwwdIChdSg3XTLZyL2fZpnKGAFC8Mc8onIQGx6OBfSlCViRJhUYDxVA4EQAALAoZAcGQ%2BioMdIhUGwADWVBUBAvAIFQUa8P0cj3TsJ7YDM%2F78sB%2BWKFJelXr%2Bd4Piu54QxO37QwlgGVSYgqOWjxiCgIeDGTsxhcW6hAAKrwPAZAMO8bACOp8JjgK%2FkKrJUkwWhhBIkAA%3D)

**Reply by jsvgoncalves** [reply]
For reference, here the current pydantic for [`ConceptScheme`](https://github.com/sentier-dev/glossary-api/blob/main/dds_glossary/model.py#L100) and [`Concept`](https://github.com/sentier-dev/glossary-api/blob/main/dds_glossary/model.py#L319)
